module-bsp/board/rt1051/bsp/magnetometer/magnetometer.cpp

```diff
@@ -118,234 +125,256 @@ namespace bsp
-                }
-                if (measurements.X > X_upper_threshold) {
-                    if (measurements.Y > Y_threshold) {
-                        position = bsp::KeyCodes::SSwitchUp;
-                    }
-                }
-                if (measurements.Y < Y_threshold) {
-                    position = bsp::KeyCodes::SSwitchMid;
-                }
-                last_slider_position = position;
+        if (measurementsMsb.int_flag) {
+            LOG_DEBUG("INT flag in register is set");
+        }
+
+        if (measurementsMsb.new_data_flag != als31300::MEAS_REG_NEW_DATA_available) {
+            return std::nullopt;
+        }
+
+        if (measurementsMsb.int_flag) {
+            if (!i2cWrite(als31300::MEASUREMENTS_MSB_REG, measurementsMsb)) {
+                LOG_ERROR("Failed to clear INT flag in register");
+                return std::nullopt;
             }
-            return position;
         }
 
-        void resetCurrentParsedValue()
-        {
-            current_parsed = bsp::KeyCodes::Undefined;
+        if (!i2cRead(als31300::MEASUREMENTS_LSB_REG, readRegister)) {
+            LOG_ERROR("Failed to read measurements LSB register!");
+            return std::nullopt;
         }
+        const als31300::measurements_LSB_reg measurementsLsb = readRegister;
 
-        std::optional<bsp::KeyCodes> WorkerEventHandler()
-        {
-            // try to get new data from active magneto
-            setActive(als31300::PWR_REG_SLEEP_MODE::active);
-            const auto new_measurement = getMeasurement();
-            setActive(als31300::PWR_REG_SLEEP_MODE::sleep);
-            if (new_measurement.has_value()) {
-                auto incoming_parsed = parse(new_measurement.value());
-                if (incoming_parsed != bsp::KeyCodes::Undefined and incoming_parsed != current_parsed) {
-                    current_parsed = incoming_parsed;
-                    return current_parsed;
-                }
+        const Measurements measurement{
+            .X = als31300::measurement_sign_convert(measurementsMsb.X_MSB << 4 | measurementsLsb.X_LSB),
+            .Y = als31300::measurement_sign_convert(measurementsMsb.Y_MSB << 4 | measurementsLsb.Y_LSB),
+            .Z = als31300::measurement_sign_convert(measurementsMsb.Z_MSB << 4 | measurementsLsb.Z_LSB)};
+        return measurement;
+    }
+
+    bsp::KeyCodes parse(const Measurements &measurements)
+    {
+        // X is tri-stable
+        constexpr auto X_lower_threshold = -85;
+        constexpr auto X_upper_threshold = 80;
+        // Y is bi-stable
+        constexpr auto Y_threshold = -200;
+        // Y is used only for proofing X, so no strict thresholds
+        // Z is useless
+
+        auto position = lastSliderPosition;
+
+        if (sliderChangedPosition(measurements)) {
+            if ((measurements.X < X_lower_threshold) && (measurements.Y > Y_threshold)) {
+                position = bsp::KeyCodes::SSwitchDown;
+            }
```

Comment:
Shouldn't be an `else` here (even for clarity)?

Reply:
From the algorithm's point of view there's no difference as the conditions are mutually exclusive, but I've added it for clarity.